apps/prairielearn/src/tests/render.test.ts

```diff
@@ -0,0 +1,365 @@
+  after('shut down testing server', helperServer.after);
+
+  internallyGradedQuestions.forEach(({ relativePath, info }) => {
+    it(`should succeed for ${relativePath}`, async function () {
+      if (unsupportedQuestions.includes(relativePath)) {
+        this.skip();
+      }
+
+      await features.runWithGlobalOverrides({ 'process-questions-in-server': false }, async () => {
+        const question = {
+          options: info.options ?? {}, // Use options from info.json if available
+          directory: relativePath,
+          type: 'Freeform',
+        } as unknown as Question;
+
+        // Prepare and generate
+        const { courseIssues: prepareGenerateIssues, variant: rawVariant } = await makeVariant(
+          question,
+          course,
+          {
+            variant_seed: null,
+          },
+        );
+
+        assert.isEmpty(prepareGenerateIssues, 'Prepare/Generate courseIssues should be empty');
+
+        const variant = rawVariant as Variant;
+        variant.num_tries = 0;
+
+        // Render
+        const locals = {
+          urlPrefix: '/prefix1',
+          plainUrlPrefix: config.urlPrefix,
+          questionRenderContext: undefined,
+          ...buildQuestionUrls(
+            '/prefix2',
+            { id: 'vid', workspace_id: 'wid' } as unknown as Variant,
+            { id: 'qid' } as unknown as Question,
+            null,
+          ),
+        };
+        const {
+          courseIssues: renderIssues,
+          data: { questionHtml },
+        } = await questionModule.render(
+          {
+            question: true,
+            submissions: false,
+            answer: false,
+          },
+          variant,
+          question,
+          null /* submission */,
+          [] /* submissions */,
+          course,
+          locals,
+        );
+        assert.isEmpty(renderIssues, 'Render courseIssues should be empty');
+
+        // Validate HTML
+        await validateHtml(questionHtml);
+
+        if (!questionModule.test) {
+          assert.fail('Test function not implemented for this question module');
+        }
+
+        const {
+          data: { raw_submitted_answer, format_errors },
+        } = await questionModule.test(variant, question, course, 'correct');
+
+        const { courseIssues: parseIssues, data: parsedData } = await questionModule.parse(
+          {
+            submitted_answer: raw_submitted_answer,
+            raw_submitted_answer,
+            gradable: true,
+          },
+          variant,
+          question,
+          course,
+        );
+
+        assert.isEmpty(
+          parseIssues,
+          `Parse courseIssues should be empty for input ${JSON.stringify(
+            raw_submitted_answer,
+            undefined,
+            2,
+          )}`,
+        );
+
+        assert.deepEqual(
+          Object.keys(parsedData.format_errors ?? {}),
+          Object.keys(format_errors),
+          'Parse format_errors should be equal',
+        );
```

Comment:
This only compares keys. Can we safely compare the entire object?

We're currently only testing `test(..., 'correct')`. Are there plans to test invalid/incorrect as well? If not, we can probably simplify all this to just assert that there aren't any format errors, which should always be the case for correct submissions.

Reply:
I am not planning to test invalid/incorrect for now.

Reply:
These were simplified.